### PR TITLE
[+] Customisation Errors - Add errors in console when user create Forest customisation with syntax errors in code

### PR DESCRIPTION
**Pull request by pierreco**
with error

### CODE REVIEW

**Comment by pierreco** [comment]
[(3) Users want to see errors when creating a smart field with syntax error.](https://trello.com/c/ECGYMLjk/1459-3-users-want-to-see-errors-when-creating-a-smart-field-with-syntax-error)

File: package.json [hunk 1]
```diff
@@ -23,6 +23,7 @@
     "bcryptjs": "2.4.3",
     "bluebird": "3.3.4",
     "body-parser": "1.15.0",
+    "esprima": "^3.1.3",
```
Comment (arnaudbesnier) [inline package.json]: We try to set fixed versions of npm packages to prevent bad surprises.

File: utils/inspect-code-syntax.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        logger.error('ERROR in ' + file);
+        logger.error('Smart Collection build failed: SyntaxError: ' +
+          errors.description + ' expected { (' + errors.lineNumber +
+          ':' + errors.index +')');
+      }
+  	}
+  })
+  .then(function () {
+    return false;
+  });
```
Comment (arnaudbesnier) [inline utils/inspect-code-syntax.js]: is it necessary to return `false`? If so `.thenReturn(false);` is lighter and easier to read.

File: utils/inspect-code-syntax.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+      return esprima.parse(fileContent.toString(),
+        { tolerant: true, loc: true });
+  	} catch (errors) {
+      if (errors) {
+        logger.error('ERROR in ' + file);
+        logger.error('Smart Collection build failed: SyntaxError: ' +
+          errors.description + ' expected { (' + errors.lineNumber +
+          ':' + errors.index +')');
+      }
+  	}
```
Comment (arnaudbesnier) [inline utils/inspect-code-syntax.js]: Indentation is **very** important to help the teammates to read the code

File: index.js [hunk 4]
```diff
@@ -114,6 +115,11 @@ exports.init = function (Implementation) {
             directorySmartImplementation = path.resolve('.') + '/forest';
           }
 
+          if(directorySmartImplementation){
```
Comment (arnaudbesnier) [inline index.js]: `directorySmartImplementation` is always set, right? Then the `if` condition is not necessary I think.

File: index.js [hunk 5]
```diff
@@ -18,6 +18,7 @@ var request = require('superagent');
 var logger = require('./services/logger');
 var Integrator = require('./integrations');
 var errorHandler = require('./services/error-handler');
+var inspectCodeSyntax = require('./utils/inspect-code-syntax');
```
Comment (arnaudbesnier) [inline index.js]: Why not call this util `codeSyntaxInspector`

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,6 +1,8 @@
 # Change Log
 
 ## [Unreleased]
+### Added
+- Smart Collections - Add errors in console when user create smart field or smart collection with syntax error
```
Comment (arnaudbesnier) [inline CHANGELOG.md]: when users add any Forest customisation (Smart Fields, Smart Segments, Smart Actions,...)

File: index.js [hunk 7]
```diff
@@ -114,6 +115,11 @@ exports.init = function (Implementation) {
             directorySmartImplementation = path.resolve('.') + '/forest';
           }
 
+          if(directorySmartImplementation){
+            inspectCodeSyntax
+            .extractCodeSyntaxErrorInDirectoryFile(directorySmartImplementation);
```
Comment (arnaudbesnier) [inline index.js]: You need to indent the second line to show this is the continuation of instruction beginning at line 119.

File: utils/inspect-code-syntax.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    var file = path.join(modelsDir, filename);
+    var fileContent = fs.readFileSync(file);
+    try {
+      return esprima.parse(fileContent.toString(),
+        { tolerant: true, loc: true });
+  	} catch (errors) {
+      if (errors) {
+        logger.error('ERROR in ' + file);
+        logger.error('Smart Collection build failed: SyntaxError: ' +
+          errors.description + ' expected { (' + errors.lineNumber +
```
Comment (arnaudbesnier) [inline utils/inspect-code-syntax.js]: What is the `{` for after `expected`?

Comment (pierreco) [inline utils/inspect-code-syntax.js]: oups

File: package.json (hunk 1 above)
Comment (pierreco) [inline package.json]: done

File: utils/inspect-code-syntax.js (hunk 2 above)
Comment (pierreco) [inline utils/inspect-code-syntax.js]: Ok, yes I think it's better

File: CHANGELOG.md (hunk 6 above)
Comment (pierreco) [inline CHANGELOG.md]: done

File: index.js (hunk 7 above)
Comment (pierreco) [inline index.js]: done

File: index.js (hunk 4 above)
Comment (pierreco) [inline index.js]: ok, I added a condition in utils to verify that the directory exists

**Pull request merged**